mbedtls/src/ssl/async_io.rs

```diff
@@ -250,4 +263,42 @@ where
     }
 }
 
-// TODO: AsyncIo impl for tokio::net::UdpSocket
+/// A `tokio::net::UdpSocket` on which `connect` was successfully called.
+///
+/// Construct this type using `ConnectedAsyncUdpSocket::connect`.
+pub struct ConnectedAsyncUdpSocket {
+    socket: UdpSocket,
+}
+
+impl ConnectedAsyncUdpSocket {
+    pub async fn connect<A: tokio::net::ToSocketAddrs>(socket: UdpSocket, addr: A) -> StdResult<Self, (IoError, UdpSocket)> {
+        match socket.connect(addr).await {
+            Ok(_) => Ok(ConnectedAsyncUdpSocket { socket }),
+            Err(e) => Err((e, socket)),
+        }
+    }
+
+    pub fn into_socket(self) -> UdpSocket {
+        self.socket
+    }
+}
+
+impl AsyncRead for ConnectedAsyncUdpSocket {
```

Comment:
To reuse the code above for `TcpStream`, I choose to impl `AsyncRead` and `AsyncWrite` for `ConnectedAsyncUdpSocket`.
Although UDP does not have concept of Stream, but in API level, `AsyncRead` and `AsyncWrite` is still compatible to it.

Reply:
Well, you could make the same argument for `Read` and `Write`. I don't think we should do this. `UdpSocket` also doesn't do this.

Reply:
So, you mean I still need to define a  trait with func `recv` `send` for `AnyAsyncIo` such as UdpSocket?

Reply:
Re-added the `AsyncIo`, integrate it with `AsyncRecv` & `AsyncSend` which are simpler version of tokio's `AsyncRead` `AsyncWrite`